File: src/main/java/org/mitallast/queue/raft/discovery/ClusterDiscovery.java
package org.mitallast.queue.raft.discovery;

import com.google.common.net.HostAndPort;
import com.google.inject.Inject;
import com.typesafe.config.Config;
import javaslang.collection.HashSet;
import javaslang.collection.Set;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.mitallast.queue.transport.DiscoveryNode;
import org.mitallast.queue.transport.TransportServer;

public class ClusterDiscovery {
    private final static Logger logger = LogManager.getLogger();
    private final DiscoveryNode self;
    private final Set<DiscoveryNode> discoveryNodes;

    @Inject
    public ClusterDiscovery(Config config) {
        Config conf = config.getConfig("raft.discovery");
        self = new DiscoveryNode(conf.getString("host"), conf.getInt("port"));
        discoveryNodes = parseDiscovery(conf);
    }

    public DiscoveryNode self() {
        return self;
    }

    public Set<DiscoveryNode> discoveryNodes() {
        return discoveryNodes;
    }

    private Set<DiscoveryNode> parseDiscovery(Config config) {
        Set<DiscoveryNode> nodes = HashSet.of(self);

        if (config.hasPath("nodes")) {
            for (String hosts : config.getStringList("nodes")) {
                for (String host : hosts.split(",")) {
                    host = host.trim();
                    if (!host.isEmpty()) {
                        HostAndPort hostAndPort = HostAndPort.fromString(host);
                        nodes = nodes.add(new DiscoveryNode(
                            hostAndPort.getHostText(),
                            hostAndPort.getPortOrDefault(TransportServer.DEFAULT_PORT)
                        ));
                    }
                }
            }
        }

        logger.info("nodes: {}", nodes);
        return nodes;
    }
}

File: src/main/java/org/mitallast/queue/raft/cluster/ClusterDiscovery.kt
package org.mitallast.queue.raft.cluster

import com.google.common.net.HostAndPort
import com.google.inject.Inject
import com.typesafe.config.Config
import io.vavr.collection.HashSet
import io.vavr.collection.Set
import org.mitallast.queue.transport.DiscoveryNode
import org.mitallast.queue.transport.TransportServer

class ClusterDiscovery @Inject constructor(config: Config) {
    private val conf = config.getConfig("raft.discovery")
    val self = DiscoveryNode(conf.getString("host"), conf.getInt("port"))
    val discoveryNodes: Set<DiscoveryNode> = parseDiscovery(conf)

    private fun parseDiscovery(config: Config): Set<DiscoveryNode> {
        var nodes: Set<DiscoveryNode> = HashSet.of(self)

        if (config.hasPath("nodes")) {
            for (hosts in config.getStringList("nodes")) {
                hosts.split(",".toRegex()).dropLastWhile { it.isEmpty() }
                        .asSequence()
                        .map { token -> token.trim { it <= ' ' } }
                        .filterNot { it.isEmpty() }
                        .map { HostAndPort.fromString(it) }
                        .forEach {
                            nodes = nodes.add(DiscoveryNode(
                                    it.host,
                                    it.getPortOrDefault(TransportServer.DEFAULT_PORT)
                            ))
                        }
            }
        }
        return nodes
    }
}
